src/sst/core/elemLoader.cc

```diff
@@ -135,92 +73,114 @@ splitPath(const std::string& searchPaths)
-            // loop all the loaders in the element
-            for ( auto* loader : elempair.second ) {
-                loader->load();
+    for ( std::string& next_path : paths ) {
+        if ( verbose ) { printf("SST-DL: Searching: %s\n", next_path.c_str()); }
+
+        if ( next_path.at(next_path.size() - 1) == '/' ) {
+            sprintf(full_path, "%slib%s.so", next_path.c_str(), elemlib.c_str());
+        }
+        else {
+            sprintf(full_path, "%s/lib%s.so", next_path.c_str(), elemlib.c_str());
+        }
+
+        if ( verbose ) { printf("SST-DL: Attempting to load %s\n", full_path); }
+
+        // use a global bind policy read from environment, default to RTLD_LAZY
+        void* handle = dlopen(full_path, bindPolicy);
+
+        if ( nullptr == handle ) {
+            if ( verbose ) { printf("SST-DL: Loading failed, error: %s\n", dlerror()); }
+        }
+        else {
+            if ( verbose ) { printf("SST-DL: Load was successful.\n"); }
+
+            found_element = true;
+
+            for ( auto& libpair : ELI::LoadedLibraries::getLoaders() ) {
+                // loop all the elements in the element lib
+                for ( auto& elempair : libpair.second ) {
+                    // loop all the loaders in the element
+                    for ( auto* loader : elempair.second ) {
+                        loader->load();
+                    }
+                }
             }
+
+				// exit the search loop, we have found the library we tried to load
+				break;
         }
     }
+
+    delete[] full_path;
+
+    if ( !found_element ) { err_os << "Error: unable to find \"" << elemlib << "\" element library\n"; }
+
     return;
 }
 
-extern "C" {
-static int
-elemCB(const char* fname, void* vd)
+void
+ElemLoader::getPotentialElements(std::vector<std::string>& potential_elements)
 {
-    std::vector<std::string>* arr  = (std::vector<std::string>*)vd;
-    const char*               base = strrchr(fname, '/') + 1; /* +1 to go past the / */
-    if ( !strncmp("lib", base, 3) ) {                         /* Filter out directories and the like */
-        arr->push_back(base + 3);                             /* skip past "lib" */
-    }
-    return 0;
-}
-}
+    std::vector<std::string> paths = splitPath(searchPaths);
 
-std::vector<std::string>
-ElemLoader::getPotentialElements()
-{
-    std::vector<std::string> res;
-    lt_dlforeachfile(searchPaths.c_str(), elemCB, &res);
-    return res;
+    for ( std::string& next_path : paths ) {
+        DIR* current_dir = opendir(next_path.c_str());
+
+        if ( current_dir ) {
+            struct dirent* dir_file;
+            while ( (dir_file = readdir(current_dir)) != nullptr ) {
+                // ensure we are only processing normal files and nothing weird
+                if ( (dir_file->d_type | DT_REG) || (dir_file->d_type | DT_LNK) ) {
+                    char* current_file = new char[strlen(dir_file->d_name) + 1];
```

Comment:
Can this just  be a `std::string` I believe you could skip the strcpy https://godbolt.org/z/qjMEPbbEf and thus also skip the `delete[]`

Reply:
@nmhamster just mark this resolved if you don't want to change it.